fabtests/hpcs/src/core/core.c

```diff
@@ -0,0 +1,1476 @@
+}
+
+static int fini_completion(struct domain_state *domain_state)
+{
+	int ret;
+
+	if (domain_state->tx_cntr != NULL) {
+		ret = fi_close(&domain_state->tx_cntr->fid);
+		if (ret) {
+			hpcs_error("unable to close tx counter\n");
+			return -1;
+		}
+		domain_state->tx_cntr = NULL;
+	}
+
+	if (domain_state->rx_cntr != NULL) {
+		ret = fi_close(&domain_state->rx_cntr->fid);
+		if (ret) {
+			hpcs_error("unable to close rx counter\n");
+			return -1;
+		}
+		domain_state->rx_cntr = NULL;
+	}
+
+	if (domain_state->tx_cq != NULL) {
+		ret = fi_close(&domain_state->tx_cq->fid);
+		if (ret) {
+			hpcs_error("unable to close tx cq\n");
+			return -1;
+		}
+		domain_state->tx_cq = NULL;
+	}
+
+	if (domain_state->rx_cq != NULL) {
+		ret = fi_close(&domain_state->rx_cq->fid);
+		if (ret) {
+			hpcs_error("unable to close rx cq\n");
+			return -1;
+		}
+		domain_state->rx_cq = NULL;
+	}
+
+	return ret;
+}
+
+static int fini_domain(struct domain_state *domain_state)
+{
+	int ret = 0;
+
+	ret = fi_close(&domain_state->endpoint->fid);
+	if (ret) {
+		hpcs_error("unable to close endpoint\n");
+		return ret;
+	}
+	domain_state->endpoint = NULL;
+
+	ret = fi_close(&domain_state->av->fid);
+	if (ret) {
+		hpcs_error("unable to close address vector\n");
+		return ret;
+	}
+
+	ret = fini_completion(domain_state);
+	if (ret)
+		return ret;
+
+	ret = fi_close(&domain_state->domain->fid);
+	if (ret) {
+		hpcs_error("unable to close domain\n");
+		return ret;
+	}
+	domain_state->domain = NULL;
+
+	return 0;
+}
+
+/*
+ * Finalizes OFI.  Safely cleans up resources created by init_ofi.  Takes a
+ * "best effort" approach - goes as far as it can and returns the first error
+ * encountered.
+ */
+int fini_ofi(struct ofi_state *ofi_state)
+{
+	int ret;
+
+	ret = fini_domain(ofi_state->domain_state);
+	if (ret)
+		return ret;
+
+	ofi_state->domain_state = NULL;
+
+	ret = fi_close(&ofi_state->fabric->fid);
+	if (ret) {
+		hpcs_error("unable to close fabric\n");
+		return ret;
+	}
+
+	return 0;
+}
+
+/*
+ * HPCS uses groups of arguments separated by a "-" or "--" (which are treated
+ * as special by getopt, and cause it to stop parsing arguments).
+ *
+ * This function updates argc and argv to start at the next group of args,
+ * with argv[0] pointing to the separator.
+ */
+static void next_args(int *argc, char * const**argv)
+{
+	int i;
+	static char* empty = "--";
+
+	/* 0th element is either binary name or previous separator. */
+	for (i=1; i<*argc; i++) {
+		if (strcmp((*argv)[i], "-") == 0 || strcmp((*argv)[i], "--") == 0)
+			break;
+	}
+
+	if (i < *argc) {
+		*argc = *argc - i;
+		*argv = &(*argv)[i];
+	} else {
+		*argc = 1;
+		*argv = &empty;
+	}
+
+	/* Current option index global variable must be reset. */
+	optind = 1;
+}
+
+static int parse_arguments(int argc, char * const* argv,
+		struct arguments **arguments)
+{
+	int longopt_idx, op, ret = 0;
+	struct option longopt[] = {
+		{"prov", required_argument, 0, 'p'},
+		{"window", required_argument, 0, 'w'},
```

Comment:
a default is used, so this may not be required.

Reply:
"required_argument" here doesn't mean that the window argument has to be given on the command line (as one might reasonably expect from the name), it means that if "-w" or "--window" is present in the command line, it must be followed by a parameter.  For example, "--window 10" is fine but "--window" is not.